Add binjs_dump command which dumps multipart .binjs file tree structure

arai-a opened the pull request:
Added `binjs_dump` command which dumps the `[TREE]` section structure, along with the hex-dump for corresponding part.
currently this supports `multipart` format only.

here's output for .binjs file which contains `print(10);`:

```
Reading.
Attempting to decode as multipart.

08                      # Script {
                        # .scope
02                      # AssertedVarScope : None
                        # .directives
00                      # list (length=0) []
                        # .statements
02                      # list (length=1) [
04                      # ExpressionStatement {
                        # .expression
00                      # CallExpression {
                        # .callee
06                      # IdentifierExpression {
                        # .name
00                      # string="print"
                        # }
                        # .arguments
02                      # list (length=1) [
0a                      # LiteralNumericExpression {
                        # .value
00 00 00 00 00 00 24 40 # float=10
                        # }
                        # ]
                        # }
                        # }
                        # ]
                        # }
```

any suggestion, especially to the output format, is welcome :)

Yoric commented on crates/binjs_es6/src/io.rs:
```diff
@@ -67,10 +67,18 @@ impl<R> Deserialization<R, String> for Deserializer<R> where R: TokenReader {
     fn deserialize(&mut self) -> Result<Vec<T>, R::Error> {
         let (len, guard) = self.reader.list()?;
+        if len > 0 {
+            dump_format!(self.reader, "list (length={}) [", len);
```
If it's a debug log, could you use the macros from crate `log`?

arai-a replied:
it's a macro to print the given file format comment in the right side of binary representation,
(that handles column position and newlines internally)
which is defined in crates/binjs_io/src/io.rs
I guess I should use more descriptive name.

Yoric commented on crates/binjs_generate_library/src/lib.rs:
```diff
@@ -963,10 +970,23 @@ impl<R> Deserialization<R, Option<{name}>> for Deserializer<R> where R: TokenRea
+                            name = field.name().to_rust_identifier_case(),
+                            spec = match field_specs.get(field.name()) {
+                                Some(s) => s,
+                                None => "???"
```
Could you document that `"???"`?

arai-a replied:
I'll replace it with unwrap() call, which I learned just yesterday :)

Yoric commented on crates/binjs_io/src/io.rs:
```diff
@@ -93,8 +93,32 @@ pub trait TokenReader where Self::Error: Debug + From<::TokenReaderError>,
     /// implementations, failure to do so will raise an assertion.
     fn untagged_tuple(&mut self) -> Result<Self::UntaggedGuard, Self::Error>;
+
+    // Functions for file format dump.
```
What's a dump, exactly?

Yoric replied:
Also, it looks like it's going to make the implementation of this trait more complicated. I suspect that providing default noop implementations would be useful, what do you think?

Yoric commented on crates/binjs_io/src/io.rs:
```diff
@@ -93,8 +93,32 @@ pub trait TokenReader where Self::Error: Debug + From<::TokenReaderError>,
+
+    // Functions for file format dump.
+
+    // Enables/disables dump.
```
Nit: `///` if it's documentation.

Yoric commented on crates/binjs_generate_library/src/lib.rs:
```diff
@@ -1,3 +1,6 @@
+//#[macro_use]
+extern crate binjs_io;
```
Do you need the dependency outside of generated code?

arai-a replied:
Looks like it was unnecessary.

arai-a commented on crates/binjs_es6/src/lib.rs:
```diff
@@ -3,6 +3,7 @@
 #![feature(box_patterns)]
 #![recursion_limit="128"] // We have deeply nested data structures...
 
+#[macro_use]
```
I put the above macro inside crates/binjs_io/src/io.rs, since I cannot find any other place to put it, to make it available everywhere which needs it.
let me know if you know better place :)
(I'm rust beginner and I think I'm missing something)

Yoric replied:
If you're using it both in `binjs_generate_library` and at runtime, it should go in `binjs_shared`. If it's only runtime, `binjs_io` is good.

I should write a README.md explaining this :)

arai-a commented on crates/binjs_generate_library/Cargo.toml:
```diff
@@ -5,6 +5,7 @@ authors = ["David Teller <D.O.Teller@gmail.com>"]
 
 [dependencies]
 binjs_meta = { path = "../binjs_meta/", version = "*" }
+binjs_io = { path = "../binjs_io/", version = "*" }
```
I'll remove.

arai-a commented:
Addressed the review comments, and also added the description to crates/binjs_io/src/io.rs.
Separated the newly added methods into FileStructurePrinter trait, which has default implementation.

arai-a commented:
Fixed documentation comments, and also added structural interpretation comment for offset (used by skip)

arai-a commented:
given that the implementation will likely throw error with offset, now I feel it's better printing the offset as well.

Yoric commented on Cargo.toml:
```diff
@@ -41,6 +41,10 @@ path = "src/bin/encode.rs"
 name = "binjs_decode"
 path = "src/bin/decode.rs"
 
+[[bin]]
```
While you're at it, now that we start to have many binaries, could you add a one-line comment explaining what it does?

Yoric commented on crates/binjs_generate_library/src/lib.rs:
```diff
@@ -921,8 +921,11 @@ impl<R> Deserializer<R> where R: TokenReader {{
 
 impl<R> InnerDeserialization<R, {name}> for Deserializer<R> where R: TokenReader {{
     fn deserialize_inner(&mut self) -> Result<{name}, R::Error> where R: TokenReader {{
+        print_file_structure!(self.reader, \"{name} {{{{\");
```
Yep, that's nicer.

arai-a replied:
The macro is used only in runtime.
I'll keep it inside binjs_io.

Yoric commented on crates/binjs_io/src/multipart/read.rs:
```diff
@@ -142,11 +142,80 @@ impl Deserializer for NodeDescriptionDeserializer {
     }
 }
 
+struct DumpCursor {
```
Nit: Could you document what this is for?

Yoric commented on src/bin/dump.rs:
```diff
@@ -0,0 +1,46 @@
+    env_logger::init();
+
+    let matches = App::new("BinJS dumper")
+        .author("David Teller, <dteller@mozilla.com>,\n\
```
I'm pretty sure you're the main author :)

Yoric commented:
(you should be able to merge whenever you're ready)
